fix: reduce excessive logging when database is unreachable

dannykopping opened the pull request:
Fixes #17045

dannykopping committed:
Backoff acquiring provisioner jobs when the database is unreachable

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Checking for, and specifically handling, database unreachability in tailnet control protocol dialer

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping commented on coderd/coderd.go:
```diff
@@ -679,6 +679,10 @@ func New(options *Options) *API {
 		DERPFn:   api.DERPMap,
 		Logger:   options.Logger,
 		ClientID: uuid.New(),
+		DatabaseHealthcheckFn: func(ctx context.Context) error {
```
Self-review: I didn't want to pull the whole store into the dialer, so opted for this approach.

dannykopping replied:
Good idea 👍 will address.

dannykopping commented on coderd/tailnet_test.go:
```diff
@@ -389,9 +426,36 @@ type agentWithID struct {
 	agent.Agent
 }
 
-func setupServerTailnetAgent(t *testing.T, agentNum int, opts ...tailnettest.DERPAndStunOption) ([]agentWithID, *coderd.ServerTailnet) {
```
Self-review: had to refactor this a bit to allow existing options to be passed in or a healthcheck func. It's a bit janky but it keeps the scope of changes tight; open to other ideas.

dannykopping commented on codersdk/database.go:
```diff
@@ -0,0 +1,7 @@
+
+const DatabaseNotReachable = "database not reachable"
+
+var ErrDatabaseNotReachable = errors.New(DatabaseNotReachable)
```
Self-review: not using anywhere yet, since we can't pass _actual_ sentinel errors around in this case, but for other use-cases it will be useful. Happy to remove, though.

dannykopping commented on provisionerd/provisionerd.go:
```diff
@@ -299,7 +300,10 @@ func (p *Server) acquireLoop() {
-		p.acquireAndRunOne(client)
+		err := p.acquireAndRunOne(client)
+		if err != nil && ctx.Err() == nil { // Only log if context is not done.
+			p.opts.Logger.Debug(ctx, "retrying to acquire job", slog.F("retry_in_ms", retrier.Delay.Milliseconds()), slog.Error(err))
```
Self-review: `acquireAndRunOne` already logs its own warning - specifically the `provisionerd was unable to acquire job` one is logged when the db is unreachable - so `Debug` is what felt most appropriate to me.

dannykopping committed:
Add len checks for returned resources

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

johnstcn commented on coderd/tailnet.go:
```diff
@@ -537,13 +539,20 @@ func NewMultiAgentController(ctx context.Context, logger slog.Logger, tracer tra
 
-func (a *InmemTailnetDialer) Dial(_ context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
+func (a *InmemTailnetDialer) Dial(ctx context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
+	if a.DatabaseHealthcheckFn != nil {
+		if err := a.DatabaseHealthcheckFn(ctx); err != nil {
+			return tailnet.ControlProtocolClients{}, xerrors.Errorf("%s: %w", codersdk.DatabaseNotReachable, err)
+		}
+	}
```
I think we need to debounce this. I'm not sure how often we're going to be dialing but my gut tells me that if we call the healthcheck on each ping _synchronously_, we're going to be

1) adding latency to the dialer,
2) adding unwanted load to the database

dannykopping committed:
Merge branch 'main' of github.com:/coder/coder into dk/17045

dannykopping commented on tailnet/controllers.go:
```diff
@@ -1381,6 +1382,14 @@ func (c *Controller) Run(ctx context.Context) {
 				}
+
+				// If the database is unreachable by the control plane, there's not much we can do, so we'll just retry later.
+				if strings.Contains(err.Error(), codersdk.DatabaseNotReachable) {
```
Alas no; it's not an actual error which is plumbed through. I'm going to try a different approach though because I agree this is smelly.

dannykopping replied:
Fixed in 4b0fd6a6c68642a9ae49408a40154ef6c4b8471b
This has the disadvantage of losing the original error details, but in this case it seems justified because we really only care about `codersdk.ErrDatabaseNotReachable` and need to conditionalize on that.

dannykopping commented on provisionerd/provisionerd.go:
```diff
@@ -290,7 +291,7 @@ func (p *Server) acquireLoop() {
 	defer func() { close(p.acquireDoneCh) }()
 	ctx := p.closeContext
-	for {
+	for retrier := retry.New(10*time.Millisecond, 1*time.Second); retrier.Wait(ctx); {
```
NOTE: This is causing some flakiness currently, because jobs are timing out waiting for provisioners, now that we're introducing some backoff.

Looking into it...

dannykopping committed:
Reset retrier after each successful job acquisition

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Wrap received error into codersdk.ErrDatabaseNotReachable

This has a downside of losing the details of the received error, but in this case it seems justified since we need to conditionalize responses based on codersdk.ErrDatabaseNotReachable

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping commented on coderd/tailnet_test.go:
```diff
@@ -365,6 +365,43 @@ func TestServerTailnet_ReverseProxy(t *testing.T) {
+	// If the healthcheck fails, we have no insight into this at this level.
+	// The dial against the control plane is retried, so we wait for the context to timeout as an indication that the
+	// healthcheck is performing as expected.
+	t.Run("Failing", func(t *testing.T) {
```
This felt wrong to me; thanks for pointing this out; it was a bit undercooked.

dannykopping commented on provisionerd/provisionerd.go:
```diff
@@ -299,7 +300,10 @@ func (p *Server) acquireLoop() {
+		err := p.acquireAndRunOne(client)
+		if err != nil && ctx.Err() == nil { // Only log if context is not done.
+			p.opts.Logger.Debug(ctx, "retrying to acquire job", slog.F("retry_in_ms", retrier.Delay.Milliseconds()), slog.Error(err))
+		}
```
As I noticed myself when tests started flaking 🤦 already fixed in f92e85281333bb958e8500839c87069cdc8dbe7a

dannykopping committed:
Replace DatabaseHealthcheckFn with interface

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

dannykopping committed:
Review suggestions

Signed-off-by: Danny Kopping <dannykopping@gmail.com>

johnstcn commented on coderd/coderd.go:
```diff
@@ -675,10 +675,11 @@ func New(options *Options) *API {
+		DERPFn:              api.DERPMap,
+		Logger:              options.Logger,
+		ClientID:            uuid.New(),
+		DatabaseHealthCheck: api.Database,
```
Nice :-)

johnstcn commented on coderd/tailnet.go:
```diff
@@ -534,16 +536,28 @@ func NewMultiAgentController(ctx context.Context, logger slog.Logger, tracer tra
-func (a *InmemTailnetDialer) Dial(_ context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
+func (a *InmemTailnetDialer) Dial(ctx context.Context, _ tailnet.ResumeTokenController) (tailnet.ControlProtocolClients, error) {
+	if a.DatabaseHealthCheck != nil {
+		if _, err := a.DatabaseHealthCheck.Ping(ctx); err != nil {
+			return tailnet.ControlProtocolClients{}, xerrors.Errorf("%w: %v", codersdk.ErrDatabaseNotReachable, err)
+		}
```
Would it make sense to return this error in the `Pinger`?

dannykopping replied:
It could, but I'd rather use this approach since it's more explicit.

spikecurtis commented on coderd/tailnet_test.go:
```diff
@@ -365,6 +368,44 @@ func TestServerTailnet_ReverseProxy(t *testing.T) {
+	_, err := dialer.Dial(ctx, nil)
+
+	// Then: the error returned reflects the database has failed its healthcheck.
+	require.ErrorIs(t, err, codersdk.ErrDatabaseNotReachable)
```
👍
